WordPress/Classes/Services/ReaderTopicService.m

```diff
@@ -369,24 +375,29 @@ - (void)toggleFollowingForTag:(ReaderTagTopic *)tagTopic success:(void (^)(void)
     BOOL oldShowInMenuValue = topic.showInMenu;
 
     // Optimistically update and save
-    topic.following = !topic.following;
-    if (topic.following) {
-        topic.showInMenu = YES;
-    }
-    [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
+    [self.coreDataStack performAndSaveUsingBlock:^(NSManagedObjectContext *context) {
+        ReaderTagTopic *topicInContext = (ReaderTagTopic *)[self.coreDataStack.mainContext existingObjectWithID:tagTopic.objectID error:nil];
+        topicInContext.following = !oldFollowingValue;
+        if (topicInContext.following) {
+            topicInContext.showInMenu = YES;
+        }
+    }];
 
     // Define failure block
     void (^failureBlock)(NSError *error) = ^void(NSError *error) {
         // Revert changes on failure
-        topic.following = oldFollowingValue;
-        topic.showInMenu = oldShowInMenuValue;
-        [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
-        if (failure) {
-            failure(error);
-        }
+        [self.coreDataStack performAndSaveUsingBlock:^(NSManagedObjectContext *context) {
+            ReaderTagTopic *topicInContext = (ReaderTagTopic *)[self.coreDataStack.mainContext existingObjectWithID:tagTopic.objectID error:nil];
+            topicInContext.following = oldFollowingValue;
+            topicInContext.showInMenu = oldShowInMenuValue;
+        } completion:^{
+            if (failure) {
+                failure(error);
+            }
+        } onQueue:dispatch_get_main_queue()];
     };
 
-    if (topic.following) {
+    if (!oldFollowingValue) {
         [self followTagWithSlug:topic.slug withSuccess:success failure:failureBlock];
     } else {
         [self unfollowTag:topic withSuccess:success failure:failureBlock];
```

Comment:
I'm reviewing it relatively late in my day, but gosh it took me a while to digest this code.

This is not a critique to your work, by the way, just an observation that optimistic updates are a nice idea, but come with a lot of extra complexity.

I wonder whether they are worth it, actually.

Also, I wish we had lower level constructs to manage them, like something to say "optimistically perform `x` with `Result<Value, Error> -> Void` completion" so that the whole revert if failed and call failure callback is handled for us. But I don't know if that's possible to generalize outside the scope of a database transaction.

Definitely a conversation with a wider scope than this PR, but wanted to drop the idea here in the meantime.

Reply:
I have seen quite a few places using this pattern, all implemented with a little bit difference.

> I wonder whether they are worth it, actually.

My understanding is, this is driven by the app's UX. For example, when tap the "Following" button on a site, the app immediately changes the button to a followed state, instead of waiting for the API to return. Other examples may include, replying to a post or comment. IMO, it's a good UX, but since the view state is tied to the Core Data model, we ended up having this optimistically saving and revert upon API failure dances all over the places.

> Also, I wish we had lower level constructs to manage them

Indeed, that would be nice. Core Data supports [this `UndoManager` thing](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506663-undomanager), which might help. I had a quick look not long ago, but that's another API in iOS SDK that's a bit difficult for me to understand.